### ISSUE TITLE
Rails 5 : has_and_belongs_to_many test gets ORA-01400 since primary key column "ID" not included in insert statement

### ISSUE THREAD

**Issue by yahonda** [rsim/oracle-enhanced#856]
### Steps to reproduce
```ruby
$ cd activerecord
$ ARCONN=oracle bundle exec ruby -W -w -I"lib:test" test/cases/associations/has_and_belongs_to_many_associations_test.rb -n test_has_and_belongs_to_many_in_a_namespaced_model_pointing_to_a_namespaced_model
```

### Expected behavior

This test should finish without any errors. 
since it have been running fine with Rails 4.2.6 and Oracle enhanced adapter 1.6.7.

### Actual behavior

```ruby
ActiveRecord::StatementInvalid: OCIError: ORA-01400: cannot insert NULL into ("ARUNIT"."ARTICLES_MAGAZINES"."ID"): INSERT INTO "ARTICLES_MAGAZINES" ("ARTICLE_ID", "MAGAZINE_ID") VALUES (:a1, :a2)
```

* background

Oracle enhanced adapter does not support `auto_increment` `AUTOINCREMENT` features yet, 
then it needs to insert primary key column "ID" explicitly.

This test works successfully with Rails 4.2.6 and Oracle enhanced adapter 1.6.7. 
It explicitly inserts "ID" column.

```sql
INSERT INTO "ARTICLES_MAGAZINES" ("MAGAZINE_ID", "ARTICLE_ID", "ID") VALUES (:a1, :a2, :a3)  [["magazine_id", 10000], ["article_id", 10000], ["id", 10000]]
```

However In Rails 5, insert statement does not specify primary key "ID" column which cannot be null.
I was trying to find which code to fix but unable to find it. 

### Other findings
I think there are "some" changes between Rails 4.2.6 and Rails 5.0.0.rc1 because of insert statement differences, as column order changed.

```ruby
$ ARCONNN=sqlite3 bundle exec ruby -W -w -I"lib:test" test/cases/associations/has_and_belongs_to_many_associations_test.rb -n test_has_and_belongs_to_many_in_a_namespaced_model_pointing_to_a_non_namespaced_model
```

* Rails 5.0.0.rc1 and sqlite3 - `article_id` then `magazine_id`
```
INSERT INTO "articles_magazines" ("article_id", "magazine_id") VALUES (?, ?)  [["article_id", 1], ["magazine_id", 1]]
```

* Rails 4.2.6 with sqlite3 - then `magazine_id` then `article_id` 
```
INSERT INTO "articles_magazines" ("magazine_id", "article_id") VALUES (?, ?)  [["magazine_id", 1], ["article_id", 1]]
```

### System configuration
**Rails version**:
5.0.0.rc1

**Arel version**:
7.0.0 with fix rails/arel#422 and rails/arel#430 
```ruby
gem 'arel', github: 'yahonda/arel', branch: 'v700_oracle12'
```

**Oracle enhanced adapter version**:
Latest `rails5` branch

**Ruby version**:
ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-linux]

**Comment by yahonda** [comment]
This is full output including stack trace.

```ruby
$ ARCONN=oracle bundle exec ruby -W -w -I"lib:test" test/cases/associations/has_and_belongs_to_many_associations_test.rb -n test_has_and_belongs_to_many_in_a_namespaced_model_pointing_to_a_namespaced_model
/home/yahonda/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/bundler-1.12.4/lib/bundler/rubygems_integration.rb:468: warning: method redefined; discarding old find_spec_for_exe
/home/yahonda/.rbenv/versions/2.3.1/lib/ruby/site_ruby/2.3.0/rubygems.rb:241: warning: previous definition of find_spec_for_exe was here
Using oracle
/home/yahonda/git/oracle-enhanced/lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb:148: warning: assigned but unused variable - col_type
/home/yahonda/git/oracle-enhanced/lib/active_record/connection_adapters/oracle_enhanced/schema_definitions.rb:7: warning: method redefined; discarding old aliased_types
/home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb:389: warning: previous definition of aliased_types was here
Run options: -n test_has_and_belongs_to_many_in_a_namespaced_model_pointing_to_a_namespaced_model --seed 15324

# Running:

E

Finished in 0.787983s, 1.2691 runs/s, 0.0000 assertions/s.

  1) Error:
HasAndBelongsToManyAssociationsTest#test_has_and_belongs_to_many_in_a_namespaced_model_pointing_to_a_namespaced_model:
ActiveRecord::StatementInvalid: OCIError: ORA-01400: cannot insert NULL into ("ARUNIT"."ARTICLES_MAGAZINES"."ID"): INSERT INTO "ARTICLES_MAGAZINES" ("ARTICLE_ID", "MAGAZINE_ID") VALUES (:a1, :a2)
    stmt.c:243:in oci8lib_230.so
    /home/yahonda/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/ruby-oci8-2.2.2/lib/oci8/cursor.rb:129:in `exec'
    /home/yahonda/git/oracle-enhanced/lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb:168:in `exec_update'
    /home/yahonda/git/oracle-enhanced/lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:124:in `block in exec_insert'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:566:in `block in log'
    /home/yahonda/git/rails/activesupport/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:560:in `log'
    /home/yahonda/git/oracle-enhanced/lib/active_record/connection_adapters/oracle_enhanced_adapter.rb:1244:in `log'
    /home/yahonda/git/oracle-enhanced/lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:110:in `exec_insert'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:124:in `insert'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb:14:in `insert'
    /home/yahonda/git/rails/activerecord/lib/active_record/relation.rb:65:in `insert'
    /home/yahonda/git/rails/activerecord/lib/active_record/persistence.rb:554:in `_create_record'
    /home/yahonda/git/rails/activerecord/lib/active_record/counter_cache.rb:128:in `_create_record'
    /home/yahonda/git/rails/activerecord/lib/active_record/locking/optimistic.rb:75:in `_create_record'
    /home/yahonda/git/rails/activerecord/lib/active_record/attribute_methods/dirty.rb:123:in `_create_record'
    /home/yahonda/git/rails/activerecord/lib/active_record/callbacks.rb:302:in `block in _create_record'
    /home/yahonda/git/rails/activesupport/lib/active_support/callbacks.rb:97:in `__run_callbacks__'
    /home/yahonda/git/rails/activesupport/lib/active_support/callbacks.rb:750:in `_run_create_callbacks'
    /home/yahonda/git/rails/activerecord/lib/active_record/callbacks.rb:302:in `_create_record'
    /home/yahonda/git/rails/activerecord/lib/active_record/timestamp.rb:68:in `_create_record'
    /home/yahonda/git/rails/activerecord/lib/active_record/persistence.rb:534:in `create_or_update'
    /home/yahonda/git/rails/activerecord/lib/active_record/callbacks.rb:298:in `block in create_or_update'
    /home/yahonda/git/rails/activesupport/lib/active_support/callbacks.rb:126:in `call'
    /home/yahonda/git/rails/activesupport/lib/active_support/callbacks.rb:506:in `block (2 levels) in compile'
    /home/yahonda/git/rails/activesupport/lib/active_support/callbacks.rb:455:in `call'
    /home/yahonda/git/rails/activesupport/lib/active_support/callbacks.rb:101:in `__run_callbacks__'
    /home/yahonda/git/rails/activesupport/lib/active_support/callbacks.rb:750:in `_run_save_callbacks'
    /home/yahonda/git/rails/activerecord/lib/active_record/callbacks.rb:298:in `create_or_update'
    /home/yahonda/git/rails/activerecord/lib/active_record/persistence.rb:152:in `save!'
    /home/yahonda/git/rails/activerecord/lib/active_record/validations.rb:50:in `save!'
    /home/yahonda/git/rails/activerecord/lib/active_record/attribute_methods/dirty.rb:30:in `save!'
    /home/yahonda/git/rails/activerecord/lib/active_record/transactions.rb:324:in `block in save!'
    /home/yahonda/git/rails/activerecord/lib/active_record/transactions.rb:395:in `block in with_transaction_returning_status'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:230:in `transaction'
    /home/yahonda/git/rails/activerecord/lib/active_record/transactions.rb:211:in `transaction'
    /home/yahonda/git/rails/activerecord/lib/active_record/transactions.rb:392:in `with_transaction_returning_status'
    /home/yahonda/git/rails/activerecord/lib/active_record/transactions.rb:324:in `save!'
    /home/yahonda/git/rails/activerecord/lib/active_record/suppressor.rb:45:in `save!'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/has_many_through_association.rb:89:in `save_through_record'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/has_many_through_association.rb:47:in `insert_record'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/collection_association.rb:587:in `block (2 levels) in concat_records'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/collection_association.rb:430:in `replace_on_target'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/collection_association.rb:423:in `add_to_target'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/collection_association.rb:586:in `block in concat_records'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/collection_association.rb:584:in `each'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/collection_association.rb:584:in `concat_records'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/has_many_association.rb:129:in `concat_records'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/has_many_through_association.rb:27:in `concat_records'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/collection_association.rb:188:in `block in concat'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/collection_association.rb:203:in `block in transaction'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:232:in `block in transaction'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb:189:in `within_new_transaction'
    /home/yahonda/git/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:232:in `transaction'
    /home/yahonda/git/rails/activerecord/lib/active_record/transactions.rb:211:in `transaction'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/collection_association.rb:202:in `transaction'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/collection_association.rb:188:in `concat'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/has_many_through_association.rb:21:in `concat'
    /home/yahonda/git/rails/activerecord/lib/active_record/associations/collection_proxy.rb:1007:in `<<'
    test/cases/associations/has_and_belongs_to_many_associations_test.rb:911:in `test_has_and_belongs_to_many_in_a_namespaced_model_pointing_to_a_namespaced_model'

1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
$
```

**Comment by yahonda** [comment]
Looks like https://github.com/rails/rails/commit/2df891dc triggers this behavior.

* ActiveRecord 4.2 (4-2-stable)

`pk_attribute?(name)` returns primary key name "id" based on self.class.primary_key
```ruby
[427, 436] in /home/yahonda/git/rails/activerecord/lib/active_record/attribute_methods.rb
   427:     def readonly_attribute?(name)
   428:       self.class.readonly_attributes.include?(name)
   429:     end
   430:
   431:     def pk_attribute?(name)
=> 432:       name == self.class.primary_key
   433:     end
   434:
   435:     def typecasted_attribute_value(name)
   436:       _read_attribute(name)
(byebug) self.class.primary_key
"id"
(byebug) self.class
Publisher::Magazine::HABTM_Articles(id: integer, article_id: integer, magazine_id: integer)
(byebug) self.class.method(:primary_key).source_location
["/home/yahonda/git/rails/activerecord/lib/active_record/attribute_methods/primary_key.rb", 72]
(byebug)
```

```ruby
/home/yahonda/git/rails/activerecord/lib/active_record/attribute_methods/primary_key.rb
62fd1d37 (Joshua Peek            2009-07-30 17:49:14 -0500  72)         def primary_key
4e380828 (Julius de Bruijn       2011-11-30 17:56:16 +0100  73)           @primary_key = reset_primary_key unless defined? @primary_key
4e380828 (Julius de Bruijn       2011-11-30 17:56:16 +0100  74)           @primary_key
62fd1d37 (Joshua Peek            2009-07-30 17:49:14 -0500  75)         end
```

* ActiveRecord 5.0 (5-0-stable)

`pk_attribute?(name)` returns always false since commit 2df891dc

```ruby
[452, 461] in /home/yahonda/git/rails/activerecord/lib/active_record/attribute_methods.rb
   452:     def readonly_attribute?(name)
   453:       self.class.readonly_attributes.include?(name)
   454:     end
   455:
   456:     def pk_attribute?(name)
=> 457:       name == self.class.primary_key
   458:     end
   459:
   460:     def typecasted_attribute_value(name)
   461:       _read_attribute(name)
(byebug) self.class.primary_key
false
(byebug) self.class
Publisher::Magazine::HABTM_Articles(id: integer, article_id: integer, magazine_id: integer)
(byebug) self.class.method(:primary_key).source_location
["/home/yahonda/git/rails/activerecord/lib/active_record/associations/builder/has_and_belongs_to_many.rb", 79]
(byebug)
```

/home/yahonda/git/rails/activerecord/lib/active_record/associations/builder/has_and_belongs_to_many.rb
```ruby
=============================================================================================
2df891dc (Aaron Patterson        2016-02-19 15:31:56 -0800  79)         def self.primary_key
2df891dc (Aaron Patterson        2016-02-19 15:31:56 -0800  80)           false
2df891dc (Aaron Patterson        2016-02-19 15:31:56 -0800  81)         end
=============================================================================================
```

**Comment by yahonda** [comment]
Closed as rails/rails#25388 closed by fix rails/rails#25578
